File: mobile/packages/footprint-expo/src/utils/send-sdk-args.ts
import type { VerifyProps } from '../types';
import { API_BASE_URL, SDK_KIND, SDK_NAME, SDK_VERSION } from './constants';
import transformKeys from './transform-keys';

const NUM_RETRIES = 3;

type SendSdkArgsRequest = {
  kind: string;
  data: Pick<
    VerifyProps,
    'publicKey' | 'authToken' | 'bootstrapData' | 'options' | 'l10n'
  >;
  is_components_sdk?: boolean;
};

type SendSdkArgsResponse = {
  token: string;
  expires_at: string;
};

const sendSdkArgsRecursive = async (
  payload: SendSdkArgsRequest,
  numRetries: number,
): Promise<SendSdkArgsResponse> => {
  return fetch(`${API_BASE_URL}/org/sdk_args`, {
    method: 'POST',
    headers: {
      'x-fp-client-version': `${SDK_NAME} ${SDK_VERSION}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  }).then(response => {
    if (response.ok) {
      return response.json();
    }
    if (numRetries > 0) {
      return sendSdkArgsRecursive(payload, numRetries - 1);
    }
    return undefined;
  });
};

const sendSdkArgs = async (
  data: SendSdkArgsRequest['data'],
  options?: { isComponentSdk?: boolean },
) => {
  const result = await sendSdkArgsRecursive(
    {
      kind: SDK_KIND,
      data: {
        ...transformKeys({
          publicKey: data.publicKey,
          authToken: data.authToken,
          userData: data.bootstrapData,
          options: data.options,
          l10n: data.l10n,
        }),
        is_components_sdk: options?.isComponentSdk,
      },
    },
    NUM_RETRIES,
  );

  return result ? result.token : undefined;
};

export default sendSdkArgs;

File: mobile/packages/footprint_flutter/lib/src/utils/send_sdk_args.dart
import 'dart:convert';

import 'package:footprint_flutter/src/config/constants.dart';
import 'package:footprint_flutter/src/models/internal/configuration.dart';
import 'package:http/http.dart' as http;

class SdkArgsResponse {
  final String? data;
  final String? error;

  SdkArgsResponse({this.data, this.error});

  bool get failed => data == null;
}

Future<SdkArgsResponse> _sendSdkArgsRecursive(
    Map<String, dynamic> payload, int numRetries) async {
  try {
    var response = await http.post(
      Uri.parse('$apiBaseUrl/org/sdk_args'),
      headers: {
        'x-fp-client-version': '$sdkName $sdkVersion',
        'Content-Type': 'application/json'
      },
      body: jsonEncode(payload),
    );

    if (response.statusCode == 200) {
      var data = jsonDecode(response.body);
      return SdkArgsResponse(data: data['token']);
    } else {
      throw Exception('Failed to fetch token');
    }
  } catch (e) {
    if (numRetries > 0) {
      return _sendSdkArgsRecursive(payload, numRetries - 1);
    }
    return SdkArgsResponse(error: e.toString());
  }
}

Future<SdkArgsResponse> sendSdkArgs(Configuration data,
    {bool? isComponentSdk}) async {
  final dataJson = data.toJson();
  if (isComponentSdk != null) {
    dataJson['is_components_sdk'] = isComponentSdk;
  }
  return await _sendSdkArgsRecursive({
    'kind': sdkKind,
    'data': dataJson,
  }, numRetries);
}
